src/v/datalake/catalog_schema_manager.cc

```diff
@@ -263,20 +267,29 @@ catalog_schema_manager::get_ids_from_table_meta(
   const iceberg::table_identifier& table_id,
   const iceberg::table_metadata& table_meta,
   iceberg::struct_type& dest_type) {
-    auto schema_iter = std::ranges::find(
-      table_meta.schemas,
-      table_meta.current_schema_id,
-      &iceberg::schema::schema_id);
-    if (schema_iter == table_meta.schemas.end()) {
+    const auto* schema = table_meta.get_schema(dest_type);
+    if (schema != nullptr) {
+        vlog(
+          datalake_log.debug,
+          "Found exact match schema: ID {}",
+          schema->schema_id);
+        return true;
+    }
+
+    schema = table_meta.get_schema(table_meta.current_schema_id);
+    if (schema == nullptr) {
         vlog(
           datalake_log.error,
           "Cannot find current schema {} in table {}",
           table_meta.current_schema_id,
           table_id);
         return errc::failed;
     }
-    auto compat_res = check_schema_compat(
-      dest_type, schema_iter->schema_struct);
+
+    // Take a copy of the source type to avoid polluting cached table metadata
+    // with compat annotations.
+    auto source_copy = schema->schema_struct.copy();
+    auto compat_res = check_schema_compat(dest_type, schema->schema_struct);
```

Comment:
Not necessarily related to this PR, but this seems like a really easy footgun to hit. If the solution in general is to make a copy of the struct beforehand, should we make `check_schema_compat` take the source schema as non-const?

Reply:
yeah good point. perhaps it would be most clear to pass by value.